src/main/java/meteordevelopment/meteorclient/utils/player/DamageUtils.java

```diff
@@ -180,95 +179,107 @@ private static double getDamageForDifficulty(double damage) {
+            double endY = box.maxY;
+            double endZ = box.maxZ + zOffset;
 
-                        if (raycast(raycastContext, obsidianPos, ignoreTerrain).getType() == HitResult.Type.MISS) i++;
+            Vec3d position = new Vec3d(0, 0, 0);
 
-                        j++;
+            for (double x = startX; x <= endX; x += xStep) {
+                for (double y = startY; y <= endY; y += yStep) {
+                    for (double z = startZ; z <= endZ; z += zStep) {
+                        ((IVec3d) position).set(x, y, z);
+
+                        if (raycast(position, source, obsidianPos, ignoreTerrain) == HitResult.Type.MISS) misses++;
+
+                        hits++;
                     }
                 }
             }
 
-            return (double) i / j;
+            return (double) misses / hits;
         }
 
         return 0;
     }
 
-    private static BlockHitResult raycast(RaycastContext context, BlockPos obsidianPos, boolean ignoreTerrain) {
-        return BlockView.raycast(context.getStart(), context.getEnd(), context, (raycastContext, blockPos) -> {
+    private static HitResult.Type raycast(Vec3d start, Vec3d end, BlockPos obsidianPos, boolean ignoreTerrain) {
+        return BlockView.raycast(start, end, null, (_null, blockPos) -> {
             BlockState blockState;
             if (blockPos.equals(obsidianPos)) blockState = Blocks.OBSIDIAN.getDefaultState();
             else {
                 blockState = mc.world.getBlockState(blockPos);
-                if (blockState.getBlock().getBlastResistance() < 600 && ignoreTerrain) blockState = Blocks.AIR.getDefaultState();
+                if (blockState.getBlock().getBlastResistance() < 600 && ignoreTerrain) return null;
             }
 
-            Vec3d vec3d = raycastContext.getStart();
-            Vec3d vec3d2 = raycastContext.getEnd();
-
-            VoxelShape voxelShape = raycastContext.getBlockShape(blockState, mc.world, blockPos);
-            BlockHitResult blockHitResult = mc.world.raycastBlock(vec3d, vec3d2, blockPos, voxelShape, blockState);
-            VoxelShape voxelShape2 = VoxelShapes.empty();
-            BlockHitResult blockHitResult2 = voxelShape2.raycast(vec3d, vec3d2, blockPos);
-
-            double d = blockHitResult == null ? Double.MAX_VALUE : raycastContext.getStart().squaredDistanceTo(blockHitResult.getPos());
-            double e = blockHitResult2 == null ? Double.MAX_VALUE : raycastContext.getStart().squaredDistanceTo(blockHitResult2.getPos());
-
-            return d <= e ? blockHitResult : blockHitResult2;
-        }, (raycastContext) -> {
-            Vec3d vec3d = raycastContext.getStart().subtract(raycastContext.getEnd());
-            return BlockHitResult.createMissed(raycastContext.getEnd(), Direction.getFacing(vec3d.x, vec3d.y, vec3d.z), BlockPos.ofFloored(raycastContext.getEnd()));
-        });
+            BlockHitResult hitResult = blockState.getOutlineShape(mc.world, blockPos).raycast(start, end, blockPos);
```

Comment:
blockState.getCollisionShape needs be used here instead of getOutlineShape, because otherwise blocks like signs, which have an outline shape but no collision shape, will result in false negatives when ignore terrain is turned off

Reply:
good catch!